Content.Client/Fluids/PuddleVisualizerSystem.cs

```diff
@@ -6,105 +6,104 @@
+    {
+        if (args.Sprite == null)
         {
-            sprite.LayerSetState(0, "sparkles", "Fluids/wet_floor_sparkles.rsi");
-            sprite.Color = sprite.Color.WithAlpha(alpha);
-            sprite.LayerSetAutoAnimated(0, false);
-            sprite.LayerSetAutoAnimated(0, true); //fixes a bug where the sparkle effect would sometimes freeze on a single frame.
+            Logger.Warning($"Missing SpriteComponent for PuddleVisualizerSystem on entityUid = {uid}");
+            return;
         }
 
-        private void EndWetFloorEffect(SpriteComponent sprite, RSI? originalRSI)
+        if (!AppearanceSystem.TryGetData(uid, PuddleVisuals.VolumeScale, out float volumeScale, args.Component)
+            || !AppearanceSystem.TryGetData(uid, PuddleVisuals.CurrentVolume, out FixedPoint2 currentVolume, args.Component)
+            || !AppearanceSystem.TryGetData(uid, PuddleVisuals.SolutionColor, out Color solutionColor, args.Component)
+            || !AppearanceSystem.TryGetData(uid, PuddleVisuals.IsEvaporatingVisual, out bool isEvaporating, args.Component))
         {
-            RandomizeState(sprite, originalRSI);
-            sprite.LayerSetAutoAnimated(0, false);
+            return;
         }
 
-        private void RandomizeState(SpriteComponent sprite, RSI? rsi)
-        {
-            var maxStates = rsi?.ToArray();
-            if (maxStates is not { Length: > 0 }) return;
+        // volumeScale is our opacity based on level of fullness to overflow. The lower bound is hard-capped for visibility reasons.
+        var cappedScale = Math.Min(1.0f, volumeScale * 0.75f + 0.25f);
 
-            var selectedState = _random.Next(0, maxStates.Length - 1); //randomly select an index for which RSI state to use.
-            sprite.LayerSetState(0, maxStates[selectedState].StateId, rsi); // sets the sprite's state via our randomly selected index.
-        }
+        var newColor = component.Recolor ? solutionColor.WithAlpha(cappedScale) : args.Sprite.Color.WithAlpha(cappedScale);
 
-        private void RandomizeRotation(SpriteComponent sprite)
+        args.Sprite.LayerSetColor(0, newColor);
+
+        // Don't consider wet floor effects if we're using a custom sprite.
+        if (component.CustomPuddleSprite)
+            return;
+
+        if (isEvaporating && currentVolume <= component.WetFloorEffectThreshold)
         {
-            float rotationDegrees = _random.Next(0, 359); // randomly select a rotation for our puddle sprite.
-            sprite.Rotation = Angle.FromDegrees(rotationDegrees); // sets the sprite's rotation to the one we randomly selected.
+            // If we need the effect but don't already have it - start it
+            if (args.Sprite.LayerGetState(0) != "sparkles")
+                StartWetFloorEffect(args.Sprite, component.WetFloorEffectAlpha);
         }
+        else
+        {
+            // If we have the effect but don't need it - end it
+            if (args.Sprite.LayerGetState(0) == "sparkles")
```

Comment:
for a proper cleanup, map this state on the prototype and then get the layer here